src/goto-programs/interpreter_evaluate.cpp

```diff
@@ -923,102 +865,101 @@ void interpretert::evaluate(
         // we fail
       }
       else if(!unbounded_size(expr.type()))
       {
+        mp_vectort dest;
         dest.resize(numeric_cast_v<std::size_t>(get_size(expr.type())));
         read(address, dest);
-        return;
+        return dest;
       }
       else
       {
+        mp_vectort dest;
         read_unbounded(address, dest);
-        return;
+        return dest;
       }
     }
   }
   else if(expr.id()==ID_typecast)
   {
-    mp_vectort tmp;
-    evaluate(to_typecast_expr(expr).op(), tmp);
+    mp_vectort tmp = evaluate(to_typecast_expr(expr).op());
 
     if(tmp.size()==1)
     {
       const mp_integer &value=tmp.front();
 
       if(expr.type().id()==ID_pointer)
       {
-        dest.push_back(value);
-        return;
+        return {value};
       }
       else if(expr.type().id()==ID_signedbv)
       {
         const auto width = to_signedbv_type(expr.type()).get_width();
         const auto s = integer2bvrep(value, width);
-        dest.push_back(bvrep2integer(s, width, true));
-        return;
+        return {bvrep2integer(s, width, true)};
       }
       else if(expr.type().id()==ID_unsignedbv)
       {
         const auto width = to_unsignedbv_type(expr.type()).get_width();
         const auto s = integer2bvrep(value, width);
-        dest.push_back(bvrep2integer(s, width, false));
-        return;
+        return {bvrep2integer(s, width, false)};
       }
       else if((expr.type().id()==ID_bool) || (expr.type().id()==ID_c_bool))
-      {
-        dest.push_back(value!=0);
-        return;
-      }
+        return {value != 0};
     }
   }
   else if(expr.id()==ID_array)
   {
+    mp_vectort dest;
     forall_operands(it, expr)
     {
-      evaluate(*it, dest);
+      mp_vectort tmp = evaluate(*it);
+      dest.insert(dest.end(), tmp.begin(), tmp.end());
     }
-    return;
+    return dest;
   }
   else if(expr.id()==ID_array_of)
   {
     const auto &ty=to_array_type(expr.type());
-    std::vector<mp_integer> size;
+
+    mp_vectort size;
     if(ty.size().id()==ID_infinity)
       size.push_back(0);
     else
-      evaluate(ty.size(), size);
+      size = evaluate(ty.size());
 
     if(size.size()==1)
     {
       std::size_t size_int = numeric_cast_v<std::size_t>(size[0]);
+      mp_vectort tmp = evaluate(to_array_of_expr(expr).op());
+      mp_vectort dest;
       for(std::size_t i=0; i<size_int; ++i)
-        evaluate(to_array_of_expr(expr).op(), dest);
-      return;
+        dest.insert(dest.end(), tmp.begin(), tmp.end());
+      return dest;
     }
   }
   else if(expr.id()==ID_with)
   {
     const auto &wexpr=to_with_expr(expr);
-    evaluate(wexpr.old(), dest);
-    std::vector<mp_integer> where;
-    std::vector<mp_integer> new_value;
-    evaluate(wexpr.where(), where);
-    evaluate(wexpr.new_value(), new_value);
+    mp_vectort dest = evaluate(wexpr.old());
+    std::vector<mp_integer> where = evaluate(wexpr.where());
+    std::vector<mp_integer> new_value = evaluate(wexpr.new_value());
```

Comment:
For consistency: shouldn't these also use `mp_vectort` instead of `std::vector<mp_integer>`?